### PR TITLE
Fix components/brave_ads layering violations

### PR DESCRIPTION
**Pull request by aseren**
Resolves https://github.com/brave/brave-browser/issues/17628

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/brave_ads/ads_service_impl.cc [hunk 1]
```diff
@@ -124,12 +121,9 @@ const unsigned int kRetriesCountOnNetworkChange = 1;
...
 constexpr char kAdNotificationUrlPrefix[] = "https://www.brave.com/ads/?";
 
-static std::map<std::string, int> g_schema_resource_ids = {
-    {ads::g_catalog_schema_resource_id, IDR_ADS_CATALOG_SCHEMA}};
-
 int GetSchemaResourceId(const std::string& name) {
-  if (g_schema_resource_ids.find(name) != g_schema_resource_ids.end()) {
-    return g_schema_resource_ids[name];
+  if (name == ads::g_catalog_schema_resource_id) {
+    return IDR_ADS_CATALOG_SCHEMA;
```
Comment (mariospr) [inline browser/brave_ads/ads_service_impl.cc:126]: :ok_hand:

**Review by mariospr — APPROVED** [review]
LGTM!!

**Review by emerick — APPROVED** [review]

**Review by bsclifton — APPROVED** [review]
I love these changes 🥰

File: browser/brave_ads/ads_service_impl.cc [hunk 2]
```diff
@@ -1514,18 +1508,19 @@ bool AdsServiceImpl::MigratePrefs(const int source_version,
...
-      {{3, 4}, &AdsServiceImpl::MigratePrefsVersion3To4},
-      {{4, 5}, &AdsServiceImpl::MigratePrefsVersion4To5},
-      {{5, 6}, &AdsServiceImpl::MigratePrefsVersion5To6},
-      {{6, 7}, &AdsServiceImpl::MigratePrefsVersion6To7},
-      {{7, 8}, &AdsServiceImpl::MigratePrefsVersion7To8},
-      {{8, 9}, &AdsServiceImpl::MigratePrefsVersion8To9},
-      {{9, 10}, &AdsServiceImpl::MigratePrefsVersion9To10},
-      {{10, 11}, &AdsServiceImpl::MigratePrefsVersion10To11}};
+  static base::NoDestructor<
+      std::map<std::pair<int, int>, void (AdsServiceImpl::*)()>>
```
Comment (goodov) [inline browser/brave_ads/ads_service_impl.cc:1512]: `base::flat_map` or `base::fixed_flat_map` would be better.

File: browser/brave_ads/ads_service_impl.cc [hunk 3]
```diff
@@ -51,7 +51,6 @@
 #include "brave/components/brave_ads/browser/ads_p2a.h"
 #include "brave/components/brave_ads/browser/ads_storage_cleanup.h"
 #include "brave/components/brave_ads/browser/frequency_capping_helper.h"
-#include "brave/components/brave_ads/browser/notification_helper.h"
 #include "brave/components/brave_ads/browser/service_sandbox_type.h"
```
Comment (goodov) [inline browser/brave_ads/ads_service_impl.cc:54]: how about moving this into `browser` and dropping `content/public/browser` dep from `component`?

File: components/brave_ads/browser/DEPS [hunk 4]
```diff
@@ -4,67 +4,4 @@ include_rules = [
   "-bat/ads/internal",
   "+bat/ledger/mojom_structs.h",
```
Comment (goodov) [inline components/brave_ads/browser/DEPS:5]: can you also check what's left here. maybe some of these are not needed anymore?

**Comment by aseren** [comment]
Resolved conflicts with chromium 96

**Review by tmancey — APPROVED** [review]
LGTM

**Review by goodov — APPROVED** [review]

**Pull request merged**